cores/esp8266/core_esp8266_wiring_digital.cpp

```diff
@@ -29,220 +29,301 @@
-    if (interrupt_reg)
-      ETS_GPIO_INTR_ENABLE();
-  }
-}
-
-void initPins() {
-  //Disable UART interrupts
-  system_set_os_print(0);
-  U0IE = 0;
-  U1IE = 0;
-
-  for (int i = 0; i <= 5; ++i) {
-    pinMode(i, INPUT);
-  }
-  // pins 6-11 are used for the SPI flash interface
-  for (int i = 12; i <= 16; ++i) {
-    pinMode(i, INPUT);
-  }
-}
-
-extern void pinMode(uint8_t pin, uint8_t mode) __attribute__ ((weak, alias("__pinMode")));
-extern void digitalWrite(uint8_t pin, uint8_t val) __attribute__ ((weak, alias("__digitalWrite")));
-extern int digitalRead(uint8_t pin) __attribute__ ((weak, alias("__digitalRead")));
-extern void attachInterrupt(uint8_t pin, voidFuncPtr handler, int mode) __attribute__ ((weak, alias("__attachInterrupt")));
-extern void detachInterrupt(uint8_t pin) __attribute__ ((weak, alias("__detachInterrupt")));
+
+    extern void cleanupFunctional(void* arg);
+
+    extern void __attachInterruptFunctionalArg(uint8_t pin, voidFuncPtrArg userFunc, void* arg, int mode, bool functional)
+    {
+        // #5780
+        // https://github.com/esp8266/esp8266-wiki/wiki/Memory-Map
+        if ((uint32_t)userFunc >= 0x40200000)
+        {
+            // ISR not in IRAM
+            ::printf((PGM_P)F("ISR not in IRAM!\r\n"));
+            abort();
+        }
+
+        if (pin < 16)
+        {
+            ETS_GPIO_INTR_DISABLE();
+            interrupt_handler_t* handler = &interrupt_handlers[pin];
+            handler->mode = mode;
+            handler->fn = (voidFuncPtr)userFunc;
+            if (handler->functional && handler->arg)  // Clean when new attach without detach
+            {
+                cleanupFunctional(handler->arg);
+            }
+            handler->arg = arg;
+            handler->functional = functional;
+            interrupt_reg |= (1 << pin);
+            GPC(pin) &= ~(0xF << GPCI);//INT mode disabled
+            GPIEC = (1 << pin); //Clear Interrupt for this pin
+            GPC(pin) |= ((mode & 0xF) << GPCI);//INT mode "mode"
+            ETS_GPIO_INTR_ATTACH(interrupt_handler, &interrupt_reg);
+            ETS_GPIO_INTR_ENABLE();
+        }
+    }
+
+    extern void __attachInterruptArg(uint8_t pin, voidFuncPtrArg userFunc, void* arg, int mode)
+    {
+        __attachInterruptFunctionalArg(pin, userFunc, arg, mode, false);
+    }
+
+    extern void __attachInterrupt(uint8_t pin, voidFuncPtr userFunc, int mode)
+    {
+        __attachInterruptFunctionalArg(pin, (voidFuncPtrArg)userFunc, 0, mode, false);
+    }
+
+    extern void __detachInterrupt(uint8_t pin)
+    {
+        if (pin < 16)
+        {
+            ETS_GPIO_INTR_DISABLE();
+            GPC(pin) &= ~(0xF << GPCI);//INT mode disabled
+            GPIEC = (1 << pin); //Clear Interrupt for this pin
+            interrupt_reg &= ~(1 << pin);
+            interrupt_handler_t* handler = &interrupt_handlers[pin];
+            handler->mode = 0;
+            handler->fn = 0;
+            if (handler->functional && handler->arg)
```

Comment:
If handler->arg was previously set via __attachInterruptArg(uint8_t pin, voidFuncPtr userFunc, void *arg, int mode), the missing functional guard would always call cleanupFunctional even on non-ArgStructure objects, which is a bug.